riscv/processor.cc

```diff
@@ -922,11 +942,25 @@ void processor_t::set_csr(int which, reg_t val)
       VU.vxsat = (val & VCSR_VXSAT) >> VCSR_VXSAT_SHIFT;
       VU.vxrm = (val & VCSR_VXRM) >> VCSR_VXRM_SHIFT;
       break;
-    case CSR_MSTATUS: {
+    case CSR_MSTATUSH:
+      {
+	reg_t mask = MSTATUS_SBE | MSTATUS_MBE;
+	if (xlen != 32)
+	  break;
+	val = (state.mstatus & ~mask) | ((val << 32) & mask);
+      }
+      if (0) {
```

Comment:
I'm not sure I understand the purpose of this, but it's nasty regardless

Reply:
The purpose is that in 32-bit mode writes to mstatus/mstatush should take either the upper or the lower 32 bits from the existing value of state.mstatus (which is always 64 bits), and then fall through to the code which handles 64 bit mode.  Thus the `if(0)` is needed to prevent the fallthrough from executing the first bit of the `CSR_MSTATUS` case, which is for writing mstatus in 32-bit mode.

Reply:
OK I understand, but this is still nasty to have if() and case intermixed like this. The language allows it, but I'd say that's a bug in the language.

Reply:
So what would you prefer?   Duplicating the logic from MSTATUS (but removing the handling of changes that can't happen from writing MSTATUSH only)?  Breaking out the code from MSTATUS into a method and calling it from both cases?